app/controllers/supervision/AnotherBodyController.scala

```diff
@@ -47,16 +50,40 @@ class AnotherBodyController @Inject() (val dataCacheConnector: DataCacheConnecto
       Form2[AnotherBody](request.body) match {
         case f: InvalidForm =>
           Future.successful(BadRequest(another_body(f, edit)))
-        case ValidForm(_, data) =>
+        case ValidForm(_, data: AnotherBody) =>
           for {
             supervision <- dataCacheConnector.fetch[Supervision](Supervision.key)
-            _ <- dataCacheConnector.save[Supervision](Supervision.key,
-              supervision.anotherBody(data)
-            )
-          } yield edit match {
-            case true => Redirect(routes.SummaryController.get())
-            case false => Redirect(routes.ProfessionalBodyMemberController.get())
+            cache <- dataCacheConnector.save[Supervision](Supervision.key,
+              updateData(supervision, data))
+          } yield {
+            redirectTo(edit, cache)
           }
       }
   }
+
+  private def updateData(supervision: Supervision, data: AnotherBody): Supervision = {
+    def updatedAnotherBody = (supervision.anotherBody, data) match {
+      case (_, d) if d.equals(AnotherBodyNo) => AnotherBodyNo
+      case (Some(ab), d:AnotherBodyYes) if ab.equals(AnotherBodyNo) => AnotherBodyYes(d.supervisorName, None, None, None)
+      case (Some(ab), d:AnotherBodyYes) => ab.asInstanceOf[AnotherBodyYes].supervisorName(d.supervisorName)
+      case (None, d:AnotherBodyYes) => AnotherBodyYes(d.supervisorName, None, None, None)
+
+    }
+    supervision.anotherBody(updatedAnotherBody).copy(hasAccepted = true)
+  }
+
+  private def redirectTo(edit: Boolean, cache: CacheMap)(implicit authContext: AuthContext, headerCarrier: HeaderCarrier) = {
+
+    import utils.ControllerHelper.{anotherBodyComplete, isAnotherBodyYes}
+
+    val anotherBody = anotherBodyComplete(cache)
+
+    def supervisionComplete = cache.getEntry[Supervision](Supervision.key).get.isComplete
```

Comment:
@slavvco - This get could be dangerous. Can we be sure we have a supervision cache entry here? Do we need to hand where we do not?

Reply:
Ok, done here and in three other places.